### PR TITLE
[FSSDK-10642] Refactor batch event processor

### PR DESCRIPTION
**Pull request by raju-opti**
## Summary
- The batch event processor has been refactored. Updates include:
    - dispatch retry
    - pending event retry logic moved from dispatcher to processor
    
## Test plan
- added new tests

## Issues
- FSSDK-10642

### CODE REVIEW

File: lib/event_processor/eventProcessor.ts [hunk 1]
```diff
@@ -13,14 +13,13 @@
...
  * limitations under the License.
  */
-// TODO change this to use Managed from js-sdk-models when available
-import { Managed } from './managed'
 import { ConversionEvent, ImpressionEvent } from './events'
 import { EventV1Request } from './eventDispatcher'
-import { EventQueue, DefaultEventQueue, SingleEventQueue, EventQueueSink } from './eventQueue'
 import { getLogger } from '../modules/logging'
 import { NOTIFICATION_TYPES } from '../utils/enums'
 import { NotificationSender } from '../core/notification_center'
```
Comment (mikechu-optimizely) [inline lib/event_processor/eventProcessor.ts:20]: Do we need these imports any longer?

File: lib/event_processor/eventProcessor.ts [hunk 2]
```diff
@@ -29,61 +28,7 @@ const logger = getLogger('EventProcessor')
```
Comment (mikechu-optimizely) [inline lib/event_processor/eventProcessor.ts:28]: Maybe we do not need `logger` any more either.

File: lib/event_processor/event_processor_factory.browser.spec.ts [hunk 3]
```diff
@@ -20,13 +20,38 @@ vi.mock('./default_dispatcher.browser', () => {
...
+  });
   return { getForwardingEventProcessor };
 });
 
-import { createForwardingEventProcessor } from './event_processor_factory.browser';
+vi.mock('./event_processor_factory', async (importOriginal) => {
+  const getBatchEventProcessor = vi.fn().mockImplementation(() => {
+    return {};
+  });
+  const original: any = await importOriginal();
```
Comment (mikechu-optimizely) [inline lib/event_processor/event_processor_factory.browser.spec.ts:33]: Do we want to take the opportunity to use `unknown` throughout?

File: lib/event_processor/event_processor_factory.node.ts [hunk 4]
```diff
@@ -17,9 +17,29 @@ import { getForwardingEventProcessor } from './forwarding_event_processor';
...
+): EventProcessor => {
+  const eventStore = options.eventStore ? getPrefixEventStore(options.eventStore) : undefined;
+  
+  return getBatchEventProcessor({
+    eventDispatcher: options.eventDispatcher || defaultEventDispatcher,
+    closingEventDispatcher: options.closingEventDispatcher,
+    flushInterval: options.flushInterval,
+    batchSize: options.batchSize,
+    retryOptions: {
+      maxRetries: 10,
```
Comment (mikechu-optimizely) [inline lib/event_processor/event_processor_factory.node.ts:40]: Can I inquire as to 10 retries for node vs 5 for other contexts?

File: lib/event_processor/forwarding_event_processor.spec.ts [hunk 5]
```diff
@@ -16,83 +16,108 @@
...
     expect(data).toEqual(makeBatchedEventV1([event]));
-    expect(notificationCenter.sendNotifications).toHaveBeenCalledOnce();
   });
 
-  it('should return a resolved promise when stop is called', async () => {
+  it('should emit dispatch event when event is dispatched', async() => {
     const dispatcher = getMockEventDispatcher();
-    const notificationCenter = getMockNotificationCenter();
-    const processor = getForwardingEventProcessor(dispatcher, notificationCenter);
+    const mockDispatch = vi.mocked(dispatcher.dispatchEvent);
```
Comment (mikechu-optimizely) [inline lib/event_processor/forwarding_event_processor.spec.ts:59]: This doesn't seem to be used. Do we need it?

File: lib/optimizely/index.ts [hunk 6]
```diff
@@ -17,7 +17,7 @@
 import { LoggerFacade, ErrorHandler } from '../modules/logging';
 import { sprintf, objectValues } from '../utils/fns';
 import { NotificationCenter } from '../core/notification_center';
-import { EventProcessor } from '../event_processor';
+import { EventProcessor } from '../event_processor/eventProcessor';
 
 import { IOdpManager } from '../core/odp/odp_manager';
 import { OdpConfig } from '../core/odp/odp_config';
```
Comment (mikechu-optimizely) [inline lib/optimizely/index.ts:23]: Maybe we can remove `OdpConfig`?

File: lib/tests/mock/create_event.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import { EventV1 } from "../../event_processor";
+import { Event } from "../../shared_types";
```
Comment (mikechu-optimizely) [inline lib/tests/mock/create_event.ts:2]: Are we going to use these later? ... or are they part of how the mocks are created in vi?

File: lib/utils/cache/async_storage_cache.react_native.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { Maybe } from "../type";
+import { AsyncCache } from "./cache";
+import AsyncStorage from '@react-native-async-storage/async-storage';
```
Comment (mikechu-optimizely) [inline lib/utils/cache/async_storage_cache.react_native.ts:3]: Cool. Are you working on [#952](https://github.com/optimizely/javascript-sdk/issues/952) here?

File: lib/utils/cache/cache.spec.ts [hunk 9]
```diff
@@ -0,0 +1,335 @@
+import { describe, it, expect, beforeEach } from 'vitest';
```
Comment (mikechu-optimizely) [inline lib/utils/cache/cache.spec.ts:1]: NIT: can we remove `beforeEach`?

File: lib/utils/executor/backoff_retry_runner.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import { resolvablePromise, ResolvablePromise } from "../promise/resolvablePromise";
+import { BackoffController } from "../repeater/repeater";
+import { AsyncFn, AsyncProducer, Fn } from "../type";
+import { scheduleMicrotask } from "../microtask";
```
Comment (mikechu-optimizely) [inline lib/utils/executor/backoff_retry_runner.ts:4]: I think we can safely remove some of the unused `import`s.

File: lib/utils/executor/backoff_retry_runner.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  task: AsyncProducer<T>, 
+  returnPromise: ResolvablePromise<T>,
+  cancelSignal: CancelSignal,
+  backoff?: BackoffController,
+  retryRemaining?: number,
+): void => {
+  task().then((res) => {
+    returnPromise.resolve(res);
+  }).catch((e) => {
+    if (retryRemaining === 0) {
```
Comment (mikechu-optimizely) [inline lib/utils/executor/backoff_retry_runner.ts:25]: NIT:
```suggestion
    if (retryRemaining <= 0) {
```

**Review by mikechu-optimizely — APPROVED** [review]
LGTM, but woof that was long.  

No showstoppers, just some thoughts 👇

File: lib/event_processor/eventProcessor.ts (hunk 1 above)
Comment (raju-opti) [reply lib/event_processor/eventProcessor.ts:20]: removed

File: lib/event_processor/eventProcessor.ts (hunk 2 above)
Comment (raju-opti) [reply lib/event_processor/eventProcessor.ts:28]: right, removed

File: lib/event_processor/event_processor_factory.browser.spec.ts (hunk 3 above)
Comment (raju-opti) [reply lib/event_processor/event_processor_factory.browser.spec.ts:33]: yeah. in this case though, we need the any as we are spreading the `original` value below

File: lib/event_processor/event_processor_factory.node.ts (hunk 4 above)
Comment (raju-opti) [reply lib/event_processor/event_processor_factory.node.ts:40]: node will most likely be used without any eventStore as there is no default store. So retrying more times to reduce the change of dropping events

Comment (raju-opti) [reply lib/event_processor/event_processor_factory.node.ts:40]: we might consider exposing maxRetries in future if needed

File: lib/utils/id_generator/index.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
+const idSuffixBase = 10_000;
+
+export class IdGenerator {
+  private idSuffixOffset = 0;
+
+  // getId returns an Id that generally increases with each call.
+  // only exceptions are when idSuffix rotates back to 0 within the same millisecond
+  // or when the clock goes back
+  getId(): string {
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/utils/id_generator/index.ts:9]: No, this id is used internally as a key for the store, it is not passed to logx

File: lib/index.node.tests.js [hunk 13]
```diff
@@ -54,17 +53,17 @@ describe('optimizelyFactory', function() {
...
-        var localLogger = loggerPlugin.createLogger({ logLevel: enums.LOG_LEVEL.INFO });
-        assert.doesNotThrow(function() {
-          var optlyInstance = optimizelyFactory.createInstance({
-            projectConfigManager: getMockProjectConfigManager(),
-            logger: localLogger,
-          });
-        });
-        sinon.assert.calledWith(localLogger.log, enums.LOG_LEVEL.ERROR);
-      });
+      // it('should not throw if the provided config is not valid and log an error if logger is passed in', function() {
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/index.node.tests.js:56]: Yeah, we will rewrite these tests competely, currently they are very hard to work with

File: lib/index.node.ts [hunk 14]
```diff
@@ -21,14 +21,13 @@ import * as loggerPlugin from './plugins/logger';
...
-import eventProcessorConfigValidator from './utils/event_processor_config_validator';
 import { createNotificationCenter } from './core/notification_center';
-import { createEventProcessor } from './plugins/event_processor';
 import { OptimizelyDecideOption, Client, Config } from './shared_types';
 import { NodeOdpManager } from './plugins/odp_manager/index.node';
 import * as commonExports from './common_exports';
 import { createPollingProjectConfigManager } from './project_config/config_manager_factory.node';
-import { createForwardingEventProcessor } from './event_processor/event_processor_factory.node';
+import { createForwardingEventProcessor, createBatchEventProcessor } from './event_processor/event_processor_factory.node';
+import { create } from 'domain';
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/index.node.ts:30]: yeah, vscode auto import. removing

File: lib/event_processor/forwarding_event_processor.spec.ts (hunk 5 above)
Comment (raju-opti) [reply lib/event_processor/forwarding_event_processor.spec.ts:59]: nope, removed

File: lib/optimizely/index.ts (hunk 6 above)
Comment (raju-opti) [reply lib/optimizely/index.ts:23]: removed

File: lib/utils/cache/async_storage_cache.react_native.ts (hunk 8 above)
Comment (raju-opti) [reply lib/utils/cache/async_storage_cache.react_native.ts:3]: Not yet. I initially tried but there is some additional complications. We'll come back to it later.

File: lib/utils/cache/cache.spec.ts (hunk 9 above)
Comment (raju-opti) [reply lib/utils/cache/cache.spec.ts:1]: removed

File: lib/utils/executor/backoff_retry_runner.spec.ts [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+    for(let i = 0; i < 500; i++) {
+      await exhaustMicrotasks();
+      await advanceTimersByTime(0);
+    }
+    expect(await result).toBe(1);
+    expect(count).toBe(500);
+  });
+
+  it('should use the backoff controller to delay retries', async () => {
+    let count = 0;
```
[the remark this answers is not in the document]
Comment (raju-opti) [reply lib/utils/executor/backoff_retry_runner.spec.ts:84]: there are other tests that tests that. Here we dont' need the count variable. Removed.

**Pull request merged**